src/sql/src/query_model/hir_generator.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use itertools::Itertools;
+
+use crate::plan::HirScalarExpr;
+use crate::query_model::{
+    BoxId, BoxType, Column, ColumnReference, Expr, Model, QuantifierType, Values,
+};
+
+use crate::plan::expr::HirRelationExpr;
+
+pub struct FromHir {
+    model: Model,
+}
+
+impl FromHir {
+    /// Generates a Query Graph Model for representing the given query.
+    pub fn generate(expr: &HirRelationExpr) -> Model {
+        let mut generator = FromHir {
+            model: Model::new(),
+        };
+        generator.model.top_box = generator.generate_select(expr);
+        generator.model
+    }
+
+    /// Generates a sub-graph representing the given expression, ensuring
+    /// that the resulting graph starts with a Select box.
+    fn generate_select(&mut self, expr: &HirRelationExpr) -> BoxId {
+        let mut box_id = self.generate_internal(expr);
+        if !self.model.get_box(box_id).borrow().is_select() {
+            box_id = self.wrap_within_select(box_id);
+        }
+        box_id
+    }
+
+    fn generate_internal(&mut self, expr: &HirRelationExpr) -> BoxId {
+        match expr {
+            // HirRelationExpr::Get { id, typ } => {
+            //     self.model.make_box(BoxType::BaseTable(BaseTable {}))
+            // }
+            HirRelationExpr::Constant { rows, typ } => {
+                if typ.arity() != 0 {
+                    panic!("expressions are not yet supported");
+                }
+                self.model.make_box(BoxType::Values(Values {
+                    rows: rows.iter().map(|_| Vec::new()).collect_vec(),
+                }))
+            }
+            HirRelationExpr::Map { input, scalars } => {
+                let box_id = self.generate_select(input);
+                // @todo self-referencing Maps
+                for scalar in scalars.iter() {
+                    let expr = self.generate_expr(scalar, box_id);
+                    let b = self.model.get_box(box_id);
+                    b.borrow_mut().columns.push(Column { expr, alias: None });
```

Comment:
We can grab the box outside of the loop, but we can't borrow it, since `generate_expr` will `borrow_mut` the box as well, for adding the subquery quantifiers. See https://github.com/MaterializeInc/materialize/pull/8775/commits/9c519e08c4cb66a2aa22f1d20ebadff2154af577

Reply:
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
  --> src/sql/src/query_model/hir_generator.rs:61:32
   |
59 |                 let b = self.model.get_box(box_id);
   |                         ---------- immutable borrow occurs here
60 |                 for scalar in scalars.iter() {
61 |                     let expr = self.generate_expr(scalar, box_id);
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
62 |                     b.borrow_mut().columns.push(Column { expr, alias: None });
   |                     - immutable borrow later used here

For more information about this error, try `rustc --explain E0502`.
error: could not compile `sql` due to previous error
```

We can't grab the box outside of the loop, since that borrows the model owned by the generator through the generator.

Reply:
Understood.